### ISSUE TITLE
Expect wrapper to be null matcher

### ISSUE THREAD

**Issue by jcarlson** [PsychoLlama/expect-enzyme#7]
I have some components that render `null` if certain conditions aren't met. Such as:

```javascript
import { shallow } from 'enzyme'
const FooBar = ({ visible }) => ( visible ? <div className='foo-bar'>Foo! Bar!</div> : null )
let rendered = shallow(<FooBar visible={false} />)
```

When rendered with Enzyme, `rendered.exists()` is `true`, but I'm looking for a convenient way to assert that it simply rendered `null`. It turns out, `rendered.equals(null)` is also true, so am suggesting a matcher along the lines of: `expect(rendered).toBeNull()`.

Thoughts?

**Comment by PsychoLlama** [comment]
@jcarlson good point! I've run into that as well. What if we overloaded `to(Not)Equal`? It should cover the `null` case and I think it would allow some sweet assertions too...

```js
expect(range).toEqual(<Range value={0.65} />)
expect(foo).toEqual(null)
expect(bacon).toNotEqual(<div>Sold out</div>)
```

Would something like that work for you?

**Comment by jcarlson** [comment]
Yeah, the only thing I would worry about is the form `expect(foo).toEqual(null)` might read as though `foo` itself should be `null`, when in fact it is a `ShallowWrapper` or a `ReactWrapper`.

Though I struggle to think of a better predicate phrase that conveys "a wrapper that renders nothing".

**Comment by PsychoLlama** [comment]
You're right, it could be confusing. Hmmm, maybe...

- `.toEqualJSX()` (might [conflict](https://github.com/algolia/expect-jsx))
- `.toMatch()` (same problem as toEqual)
- `.looksLike()` (a bit weird)
- `.toEqualEnzyme()` (a bit more weird)
- `.toHaveRendered()`
- `.toRender()`
- `.toEqualThisOtherThingButInReactWords()`

I think my favorite there is `.toRender()`, but it still doesn't quite fit. I'll give this some more thought...

**Comment by PsychoLlama** [comment]
```js
expect(foo).toLookLike(null)
expect(button).toLookLike(<Button disabled />)
expect(upload.find('Progress')).toLookLike(<ProgressBar value={100} />
```

```js
expect(foo).toHaveRendered(null)
expect(button).toHaveRendered(<Button disabled />)
expect(upload.find('Progress')).toHaveRendered(<ProgressBar value={100} />
```

Naming things is hard.

**Comment by PsychoLlama** [comment]
Quick update: I've written the assertion, it was pretty straightforward. The tricky part has been the error message. Here's what I'm headed for:

```js
expect(inputComponent).toHaveRendered(<textarea string="value" attr={true}>truncated content</textarea>)
// Expected div to render <textarea string="value" attr>...</textarea>
```

Stringifying elements is pretty straightforward, but there are some nasty gotchas. I'm currently wrangling with [`react-element-to-jsx-string`](https://github.com/algolia/react-element-to-jsx-string), but it outputs the string across multiple lines. A replace for `\n` messes up some indentation.

Anyway, that's where I'm at. Wanted you to know I hadn't forgotten about this request :)

**Comment by jcarlson** [comment]
Nice, thanks! If you haven't looked at it already, you might check out: https://github.com/grommet/jsx-to-string

We use it in combination with react-storybook in one of our projects and it seems to work pretty well.

**Comment by PsychoLlama** [comment]
Sweet, it looks like that'll solve the problem! :tada:

**Comment by PsychoLlama** [comment]
Tada! [`.toHaveRendered()`](https://github.com/PsychoLlama/expect-enzyme#tohaverenderedelement) is now on master.

I ended up writing my own element stringifier so I had more control over what values were abbreviated. If you pass an array as a prop it'll show you `Array[148]`, where 148 is the length. It shows a pretty value for react elements passed as props, functions, subclassed objects, symbols, regexes... really I just went wild for a bit. ~Still a few things to tidy up though regarding `children`~ (finished). Should have it published before Monday.

That was a fun assertion to write.
